### ISSUE TITLE
Filesystem error durring upload of file to session results in file record remaining in session collection.

### ISSUE THREAD

**Issue by ryansanford** [scitran/api#115]
@gsfr @rentzso Let me know if I can provide any additional detail.

## How to reproduce
1. chmod -R -w "${SCITRAN_PERSISTENT_DATA_PATH}"
2. upload file
```
# Headers & settings
baseUrl="https://inv-ryan-03.invenshure.com:8443/api/sessions"
alias sCurl="curl -sk -H 'User-Agent: SciTran Drone Engine' -H 'X-SciTran-Auth: change-me'"

# Get a session as an upload target
SessionId=$(sCurl $baseUrl | jq -r '.[0]._id')

# Generate a useless file
uname -a > example.bin

# Succeeds
sCurl $baseUrl/$SessionId/files -F "file=@example.bin"

```

## Error
```
2016-01-04 16:03:04      scitran.api                  base.py   127:DEBU from None None POST /api/sessions/568a8893e460bcea7c3fdd2a/files {}
2016-01-04 16:03:04      scitran.api           liststorage.py    41:DEBU query {'_id': ObjectId('568a8893e460bcea7c3fdd2a')}
2016-01-04 16:03:04      scitran.api            validators.py   125:DEBU {u'key_fields': [u'name'], u'required': [u'name', u'created', u'modified', u'size', u'hash', u'unprocessed'], u'additionalProperties': False, u'$schema': u'http://json-schema.org/draft-04/schema#', u'type': u'object', u'properties': {u'hash': {u'type': u'string'}, u'name': {u'type': u'string'}, u'created': {}, u'unprocessed': {u'type': u'boolean'}, u'measurements': {u'uniqueItems': True, u'items': {u'type': u'string'}, u'type': u'array'}, u'modified': {}, u'instrument': {u'type': u'string'}, u'metadata': {u'type': u'object'}, u'type': {u'type': u'string'}, u'tags': {u'uniqueItems': True, u'items': {u'type': u'string'}, u'type': u'array'}, u'size': {u'type': u'integer'}}}
2016-01-04 16:03:04      scitran.api            validators.py    75:DEBU {'hash': '128a2aa7f6167cb09592541c97073b2bdc6e2a89b4a39a8cfad81f476e43eaa10b844f34811154632000050bd8b782de', 'name': 'example.bin', 'created': datetime.datetime(2016, 1, 4, 16, 3, 4, 667687), 'unprocessed': True, 'modified': datetime.datetime(2016, 1, 4, 16, 3, 4, 667687), 'size': 136}
2016-01-04 16:03:04      scitran.api           liststorage.py    65:DEBU payload {'hash': '128a2aa7f6167cb09592541c97073b2bdc6e2a89b4a39a8cfad81f476e43eaa10b844f34811154632000050bd8b782de', 'name': 'example.bin', 'created': datetime.datetime(2016, 1, 4, 16, 3, 4, 667687), 'unprocessed': True, 'modified': datetime.datetime(2016, 1, 4, 16, 3, 4, 667687), 'size': 136}
2016-01-04 16:03:04      scitran.api           liststorage.py    70:DEBU query {'files': {'$not': {'$elemMatch': {u'name': 'example.bin'}}}, '_id': ObjectId('568a8893e460bcea7c3fdd2a')}
2016-01-04 16:03:04      scitran.api           liststorage.py    71:DEBU update {'$push': {'files': {'hash': '128a2aa7f6167cb09592541c97073b2bdc6e2a89b4a39a8cfad81f476e43eaa10b844f34811154632000050bd8b782de', 'name': 'example.bin', 'created': datetime.datetime(2016, 1, 4, 16, 3, 4, 667687), 'unprocessed': True, 'modified': datetime.datetime(2016, 1, 4, 16, 3, 4, 667687), 'size': 136}}}
2016-01-04 16:03:04             root               webapp2.py  1552:ERRO [Errno 13] Permission denied: '/var/scitran/data/1/2'
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 1535, in __call__
    rv = self.handle_exception(request, response, e)
  File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 1529, in __call__
    rv = self.router.dispatch(request, response)
  File "./api/api.py", line 118, in dispatcher
    rv = router.default_dispatcher(request, response)
  File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 1278, in default_dispatcher
    return route.handler_adapter(request, response)
  File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 1102, in __call__
    return handler.dispatch()
  File "./api/base.py", line 128, in dispatch
    return super(RequestHandler, self).dispatch()
  File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 572, in dispatch
    return self.handle_exception(e, self.app.debug)
  File "/usr/local/lib/python2.7/dist-packages/webapp2.py", line 570, in dispatch
    return method(*args, **kwargs)
  File "./api/handlers/listhandler.py", line 400, in post
    file_store.move_file(dest_path)
  File "./api/files.py", line 112, in move_file
    os.makedirs(target_dir)
  File "/usr/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/usr/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/usr/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/usr/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/usr/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/usr/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/usr/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/var/scitran/data/1/2'
[pid: 15|app: 0|req: 29/52] 192.168.3.177 () {40 vars in 680 bytes} [Mon Jan  4 16:03:04 2016] POST /api/sessions/568a8893e460bcea7c3fdd2a/files => generated 228 bytes in 88 msecs (HTTP/1.1 500) 2 headers in 99 bytes (2 switches on core 1)
2016-01-04 16:03:09 scitran.api.jobs                 rules.py    59:WARN file example.bin in container 568a8893e460bcea7c3fdd2a has no type key
```

**Comment by gsfr** [comment]
@rentzso: Please take this on. I'd suggest that we generally prefer orphaned files over stale database entries. That is, for uploads, place file first, then update database, and for deletes, update database first, then delete file. Actually, with our new hash-based file storage, deletion will need to be handled by a cleanup job. For that, we just need an endpoint; cron functionality will be handled external to the api.

**Comment by rentzso** [comment]
I have created this pull request to fix this:
https://github.com/scitran/api/pull/118

The code was handling and cleaning only after IOErrors, but not OSErrors.
Deletes are already done only on the DB, and not on the file system and, as @gsfr was saying, we need the endpoint to do the cleanup.
